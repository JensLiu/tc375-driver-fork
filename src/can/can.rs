// CanNode trait impl to be moved on a separate file (annabo)
pub trait ACanModule {
    fn node_id(&self) -> u8;
    fn is_enabled(&self) -> bool;
    fn is_suspended(&self) -> bool;

    fn enable_module(&self);
    fn disable_module(&self);
    fn reset_module(&self);
    fn init_module(&self);
    //fn set_clock_source(&self, clock_select: ClockSelect, clock_source: ClockSource);
}

pub struct CanModule0 {
    inner: tc37x_pac::can0::Can0,
}

// TODO (annabo) use tc37x_pac::can1::Can1; impl CanModule for Can1
use crate::scu;

#[cfg(feature = "log")]
#[cfg(target_arch = "tricore")]
use defmt::println;
use embedded_can::nb::Can;

impl ACanModule for CanModule0 {
    fn node_id(&self) -> u8 {
        0
    }

    #[inline]
    fn is_enabled(&self) -> bool {
        !unsafe { self.inner.clc().read() }.diss().get()
    }

    fn is_suspended(&self) -> bool {
        unsafe { self.inner.ocs().read() }.sussta().get()
    }
   
    fn enable_module(&self) {
        let passw = scu::wdt::get_cpu_watchdog_password();

        #[cfg(feature = "log")]
        println!("enable module watchdog passw: {:x}", passw);

        scu::wdt::clear_cpu_endinit_inline(passw);

        unsafe { self.inner.clc().modify_atomic(|r| r.disr().set(false)) };
        while !self.is_enabled() {}

        scu::wdt::set_cpu_endinit_inline(passw);
    }

    fn disable_module(&self) {
        let passw = scu::wdt::get_cpu_watchdog_password();

        scu::wdt::clear_cpu_endinit_inline(passw);
        unsafe { self.inner.clc().modify(|r| r.disr().set(true)) };

        while self.is_enabled() {}

        scu::wdt::set_cpu_endinit_inline(passw);
    }

    fn reset_module(&self) {
        let passw = scu::wdt::get_cpu_watchdog_password();

        scu::wdt::clear_cpu_endinit_inline(passw);
        unsafe { self.inner.krst0().modify(|r| r.rst().set(true)) };
        unsafe { self.inner.krst1().modify(|r| r.rst().set(true)) };
        scu::wdt::set_cpu_endinit_inline(passw);

        while !unsafe { self.inner.krst0().read() }.rststat().get() {}

        scu::wdt::clear_cpu_endinit_inline(passw);
        unsafe { self.inner.krstclr().init(|r| r.clr().set(true)) };
        scu::wdt::set_cpu_endinit_inline(passw);
    }

    // fn set_clock_source(&self, clock_select: ClockSelect, clock_source: ClockSource) {
    //     let mut mcr = unsafe { tc37x_pac::CAN0.mcr().read() };
    //     mcr = mcr.ccce().set(true).ci().set(true);

    //     unsafe { tc37x_pac::CAN0.mcr().write(mcr) };

    //     mcr = mcr.clksel(clock_select.into()).set(clock_source.into());
    //     unsafe { tc37x_pac::CAN0.mcr().write(mcr) };

    //     mcr = mcr.ccce().set(false).ci().set(false);
    //     unsafe { tc37x_pac::CAN0.mcr().write(mcr) };
    // }
    fn init_module(&self) {
        if !self.is_enabled() {
            #[cfg(feature = "log")]
            defmt::debug!("module was disabled, enabling it");
            self.enable_module();
        }
    }
}

impl CanModule0{

    pub fn new() -> Self {
        let m = Self {
            inner: tc37x_pac::CAN0,
        }; 
        m.enable_module(); 
        m
        

    }

}