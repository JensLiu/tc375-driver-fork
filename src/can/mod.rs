// TODO Remove this once the module is more complete
#![allow(dead_code)]

mod baud_rate;
mod can_module;
mod can_node;
mod frame;
mod internals;
pub mod msg;
mod reg;

pub use baud_rate::{AutoBitTiming, BitTimingConfig, DataBitTiming, NominalBitTiming};
pub use can_module::{Module, ModuleConfig, ModuleId};
pub use can_node::{DataFieldSize, Node, NodeConfig, NodeId, TxConfig, TxMode};
pub use frame::Frame;
pub use msg::MessageId;
