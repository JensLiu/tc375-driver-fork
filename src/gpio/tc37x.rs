pub use super::Input as DefaultMode;
use super::*;

gpio!(gpio00, crate::pac::port_00::Port00, 0, P00n, [
    P00_0 : (p00_0 , 0 , [  0,     1,     2,     3,   /*4,*/   5,   /*6,*/   7  ]),
    P00_1 : (p00_1 , 1 , [  0,     1,     2,   /*3,*/   4,   /*5,*/   6,     7  ]),
    P00_2 : (p00_2 , 2 , [  0,     1,     2,   /*3,*/   4,     5,     6,     7  ]),
    P00_3 : (p00_3 , 3 , [  0,     1,     2,   /*3,*/   4,   /*5,*/   6,     7  ]),
    P00_4 : (p00_4 , 4 , [  0,     1,     2,     3,     4,   /*5,*/   6,     7  ]),
    P00_5 : (p00_5 , 5 , [  0,     1,     2,   /*3,*/ /*4,*/   5,     6,     7  ]),
    P00_6 : (p00_6 , 6 , [  0,     1,     2,     3,     4,     5,     6,     7  ]),
    P00_7 : (p00_7 , 7 , [  0,     1,     2,     3,     4,     5,     6,     7  ]),
    P00_8 : (p00_8 , 8 , [  0,     1,     2,     3,   /*4,*/   5,     6,     7  ]),
    P00_9 : (p00_9 , 9 , [  0,     1,     2,     3,     4,     5,     6,     7  ]),
    P00_10: (p00_10, 10, [  0,     1,     2,   /*3,*/ /*4,*/ /*5,*/   6,     7  ]),
    P00_11: (p00_11, 11, [  0,     1,     2,   /*3,*/   4   /*5,*/ /*6,*/ /*7*/]),
    P00_12: (p00_12, 12, [  0,     1,   /*2,*/ /*3,*/ /*4,*//*5,*/ /*6,*/    7  ]),
]);

gpio!(gpio01, crate::pac::port_01::Port01, 1, P01n, [
    P01_3 : (p01_3 , 3 , [  0,     1,     2,     3,     4,     5,     6,     7  ]),
    P01_4 : (p01_4 , 4 , [  0,     1,     2,     3,     4,     5,     6,     7  ]),
    P01_5 : (p01_5 , 5 , [  0,     1,     2,     3,     4,     5,     6,     7  ]),
    P01_6 : (p01_6 , 6 , [  0,     1,     2,     3,     4,     5,     6,     7  ]),
    P01_7 : (p01_7 , 7 , [  0,     1,     2,     3,     4,     5,     6,     7  ]),
]);


