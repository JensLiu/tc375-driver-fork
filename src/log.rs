//! Logging macros and utilities.

// Log with defmt
#[cfg(all(not(feature = "log_with_env_logger"), feature = "log_with_defmt"))]
pub use tricore_log::*;

// No log system
#[cfg(all(not(feature = "log_with_defmt"), not(feature = "log_with_env_logger")))]
pub use crate::{
    no_log_debug as debug, no_log_error as error, no_log_info as info, no_log_trace as trace,
    no_log_warn as warn,
};

// Log with log crate
#[cfg(all(not(feature = "log_with_defmt"), feature = "log_with_env_logger"))]
pub use ::log::{debug, error, info, trace, warn};

// link with defmt implementation
#[cfg(feature = "log_with_defmt")]
extern crate defmt_rtt;

#[cfg(feature = "log_with_defmt")]
mod tricore_log {
    pub use defmt::debug;
    pub use defmt::error;
    pub use defmt::info;
    pub use defmt::trace;
    pub use defmt::warn;
}

#[macro_export]
macro_rules! no_log_debug {
    ($($ignored:tt)*) => {};
}

#[macro_export]
macro_rules! no_log_trace {
    ($($ignored:tt)*) => {};
}

#[macro_export]
macro_rules! no_log_info {
    ($($ignored:tt)*) => {};
}

#[macro_export]
macro_rules! no_log_warn {
    ($($ignored:tt)*) => {};
}

#[macro_export]
macro_rules! no_log_error {
    ($($ignored:tt)*) => {};
}

/// A wrapper around a slice of bytes that implements `Display` to print the
/// bytes as a hex string.
///
/// ```
/// use bw_r_drivers_tc37x::log::HexSlice;
/// let bytes = &[1u8, 2, 3, 4];
/// assert_eq!(format!("{}", HexSlice::from(bytes)), "01020304");
/// ```
pub struct HexSlice<'a>(&'a [u8]);

impl core::fmt::Display for HexSlice<'_> {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        for byte in self.0 {
            write!(f, "{:02x}", byte)?;
        }
        Ok(())
    }
}

#[cfg(feature = "log_with_defmt")]
impl defmt::Format for HexSlice<'_> {
    fn format(&self, f: defmt::Formatter) {
        for byte in self.0 {
            defmt::write!(f, "{:02x}", byte);
        }
    }
}

impl<'a, const N: usize> From<&'a [u8; N]> for HexSlice<'a> {
    fn from(value: &'a [u8; N]) -> Self {
        HexSlice(value.as_slice())
    }
}
impl<'a> From<&'a [u8]> for HexSlice<'a> {
    fn from(value: &'a [u8]) -> Self {
        HexSlice(value)
    }
}
