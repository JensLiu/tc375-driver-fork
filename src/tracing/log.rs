#![allow(clippy::expect_used)]

use super::TraceGuard;
use crate::tracing::{LoadModifyStoreEntry, ReadEntry, ReportEntry, WriteEntry};
use std::any::Any;
use std::collections::{HashMap, VecDeque};
use std::fmt::{Debug, Display, Formatter, Write};
use std::sync::{Arc, Mutex, MutexGuard};

struct ReadFifoEntry {
    addr: usize,
    len: usize,
    val: u64,
}

#[derive(Default)]
struct ReadFifo(VecDeque<ReadFifoEntry>);

#[derive(Default)]
struct SharedData {
    log: Log,
    read_fifo: ReadFifo,
}

struct Reporter {
    shared_data: Arc<Mutex<SharedData>>,
}

pub struct Report {
    shared_data: Arc<Mutex<SharedData>>,
    _guard: TraceGuard,
}

impl Default for Report {
    fn default() -> Self {
        Self::new()
    }
}

impl Report {
    pub fn new() -> Self {
        let data = Arc::new(Mutex::new(SharedData::default()));
        let reporter = Reporter {
            shared_data: Arc::clone(&data),
        };
        let guard = TraceGuard::new(reporter);
        Self {
            shared_data: data,
            _guard: guard,
        }
    }

    pub fn take_log(&self) -> Log {
        let mut g = self.shared_data();
        let len = g.log.0.len();
        Log(g.log.0.drain(0..len).collect())
    }

    pub fn expect_read(&self, addr: usize, len: usize, val: u64) {
        self.shared_data()
            .read_fifo
            .0
            .push_front(ReadFifoEntry { addr, len, val })
    }

    pub fn comment(&self, s: impl Into<String>) {
        self.shared_data().log.0.push(LogEntry::Comment(s.into()));
    }

    fn shared_data(&self) -> MutexGuard<SharedData> {
        self.shared_data.lock().unwrap()
    }
}

impl Reporter {
    fn push(&self, report: ReportEntry) {
        self.shared_data().log.0.push(LogEntry::ReportEntry(report));
    }

    fn shared_data(&self) -> MutexGuard<SharedData> {
        self.shared_data.lock().unwrap()
    }
}

impl super::Reporter for Reporter {
    fn read_volatile(&self, addr: usize, len: usize) -> u64 {
        let entry = self
            .shared_data()
            .read_fifo
            .0
            .pop_front()
            .expect("Unexpected read");

        if entry.addr == addr && entry.len == len {
            let val = entry.val;
            self.push(ReportEntry::Read(ReadEntry { addr, len, val }));
            val
        } else {
            panic!("Unexpected read at address 0x{:08X} and len {}", addr, len)
        }
    }

    fn write_volatile(&self, addr: usize, len: usize, val: u64) {
        self.push(ReportEntry::Write(WriteEntry { addr, len, val }));
    }

    fn load_modify_store(&self, addr: usize, val: u64) {
        self.push(ReportEntry::LoadModifyStore(LoadModifyStoreEntry {
            addr,
            val,
        }));
    }
}

impl Drop for Reporter {
    fn drop(&mut self) {
        if !self.shared_data().read_fifo.0.is_empty() {
            panic!("More read where expected");
        }
    }
}

#[derive(Default, Debug, PartialEq)]
pub struct Log(Vec<LogEntry>);

#[derive(Debug, PartialEq)]
enum LogEntry {
    Comment(String),
    ReportEntry(ReportEntry),
}

impl Display for Log {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        for entry in &self.0 {
            match entry {
                LogEntry::ReportEntry(ReportEntry::Read(x)) => {
                    write!(f, "r    0x{:08X} {:02} 0x{:08X}", x.addr, x.len, x.val);
                }
                LogEntry::ReportEntry(ReportEntry::Write(x)) => {
                    write!(f, "w    0x{:08X} {:02} 0x{:08X}", x.addr, x.len, x.val);
                }
                LogEntry::ReportEntry(ReportEntry::LoadModifyStore(x)) => {
                    let mask = (x.val >> 32);
                    let val = (x.val & 0xFFFFFFFF);
                    write!(f, "ldms 0x{:08X} 0x{:08X} 0x{:08X}", x.addr, mask, val);
                }
                LogEntry::Comment(s) => {
                    write!(f, "# {s}");
                }
            }

            f.write_char('\n');
        }

        Ok(())
    }
}
