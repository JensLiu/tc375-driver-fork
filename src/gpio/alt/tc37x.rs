use super::*;
use crate::gpio::{self, NoPin, OpenDrain, PushPull};

pub mod can0 {
    use super::*;

    pin! {
        <Rx, PushPull> for no:NoPin, [
            // FIXME Alternate function here?
            P20_7<1>, // CAN00:RXDB = P20.7:IN
        ],

        <Tx, PushPull> for no:NoPin, [
            P20_8<5>, // CAN00:TXD = P20.8:ALT(5)
        ],
    }

    impl CanCommon for crate::pac::can0::Can0 {
        type Rx = Rx;
        type Tx = Tx;
    }
}
