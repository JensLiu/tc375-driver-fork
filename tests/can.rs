use tc37x_driver::tracing::log::Report;
use tc37x_driver::can::{
    Module,
    Module0,
    NodeConfig,
    BitTimingConfig,
    AutoBitTiming,
    Node0,
    TxConfig,
    TxMode,
    DataFieldSize,
    RxConfig,
    RxMode,
    RxFifoMode,
    Pins,
    pin_map::{PIN_TX_0_0_P20_8, PIN_RX_0_0_P20_7},

};
use tc37x as pac;
use pac::{CAN0, SCU};

// TODO fix values of can_module.enable reads
// TODO add report comments with actual registers' name
#[test]
fn test_can_module_take_node(){
    let report = Report::new();
    let can_module = Module::new(Module0);

    // wtdcpu0con0 for clear_cpu_endinit
    report.expect_read(SCU.wdtcpu0con0().addr(), 4, 0b1111111111111100_00000000111100_1_1);
    report.expect_read(SCU.wdtcpu0con0().addr(), 4, 0b1111111111111100_00000000111100_1_1);
    report.expect_read(SCU.wdtcpu0con0().addr(), 4, 0b1111111111111100_00000000111100_1_1);
    // wtdcpu0con0 write
    // wtdcpu0con0 read
    report.expect_read(SCU.wdtcpu0con0().addr(), 4, 0b1111111111111100_00000000111100_0_1);

    // wtdcpu0con0 write

    // clc read
    report.expect_read(CAN0.clc().addr(), 4, 0b0);

    // wtdcpu0con0 for set_cpu_endinit
    report.expect_read(SCU.wdtcpu0con0().addr(), 4, 0b1111111111111100_00000000111100_1_0);
    report.expect_read(SCU.wdtcpu0con0().addr(), 4, 0b1111111111111100_00000000111100_1_0);
    report.expect_read(SCU.wdtcpu0con0().addr(), 4, 0b1111111111111100_00000000111100_1_0);

    // wtdcpu0con0 write
    report.expect_read(SCU.wdtcpu0con0().addr(), 4, 0b1111111111111100_00000000111100_0_1);
    report.expect_read(SCU.wdtcpu0con0().addr(), 4, 0b1111111111111100_00000000111100_1_1);

    let mut can_module = can_module.enable();

    let cfg = NodeConfig {
        baud_rate: BitTimingConfig::Auto(AutoBitTiming {
            baud_rate: 1_000_000,
            sample_point: 8_000,
            sync_jump_width: 3,
        }),
        ..Default::default()
    };

    // mcr read for set clock source
    report.expect_read(CAN0.mcr().addr(), 4, 0b0);
    report.expect_read(CAN0.mcr().addr(), 4, 0b11);

    // cccr for enable configuration change
    // read
    report.expect_read(CAN0.n()[0].cccri().addr(), 4, 0b1);
    
    // modify
    report.expect_read(CAN0.n()[0].cccri().addr(), 4, 0b1);

    // read
    report.expect_read(CAN0.n()[0].cccri().addr(), 4, 0b1);
    
    // modify
    report.expect_read(CAN0.n()[0].cccri().addr(), 4, 0b1);
    
    // read
    report.expect_read(CAN0.n()[0].cccri().addr(), 4, 0b0);
    
    // modify
    report.expect_read(CAN0.n()[0].cccri().addr(), 4, 0b0);
    
    // read
    report.expect_read(CAN0.n()[0].cccri().addr(), 4, 0b1);

    // modify
    report.expect_read(CAN0.n()[0].cccri().addr(), 4, 0b1);

    // ccucon1 for get_mcan_frequency for configure_baud_rate
    report.expect_read(SCU.ccucon1().addr(), 4, 0b0010_0001_0001_0001_0000_0010_0001_0010);

    // ccucon0 read for get_source_frequency for get_mcan_frequency
    report.expect_read(SCU.ccucon0().addr(), 4, 0b0001_0111_0010_0011_0000_0001_0001_0011);

    // syspllcon0 read for get_osc_frequency for get_per_pll_frequency1 for get_source_frequency
    report.expect_read(SCU.syspllcon0().addr(), 4, 0b0100_0000_0000_0001_0011_1010_0000_0000);

    // perpllcon0 read for get_per_pll_frequency1 for get_source_frequency
    report.expect_read(SCU.perpllcon0().addr(), 4, 0b10011_1111_0000_0000);

    // perpllcon1 read for get_per_pll_frequency1 for get_source_frequency
    report.expect_read(SCU.perpllcon1().addr(), 4, 0b1_0000_0001);

    // ccucon1 for get_source_frequency
    report.expect_read(SCU.ccucon1().addr(), 4, 0b10_0001_0001_0001_0000_0010_0001_0010);

    // nbtp0 for set_nominal_bit_timing
    report.expect_read(CAN0.n()[0].nbtpi().addr(), 4, 0b110_0000_0000_0000_1010_0000_0011);

    let mut node = can_module.take_node(Node0, cfg).expect("Cannot take can node");

    // txesc0 for set_tx_buffer_data_field_size for setup_tx
    report.expect_read(CAN0.n()[0].tx().txesci().addr(), 4, 0b0);

    // txbc0 for set_tx_buffer_start_address for setup_tx
    report.expect_read(CAN0.n()[0].tx().txbci().addr(), 4, 0b0);

    // txbc0 for set_dedicated_tx_buffers_number for setup_tx
    report.expect_read(CAN0.n()[0].tx().txbci().addr(), 4, 0b100_0100_0000);

    // txbtie0 for enable_tx_buffer_transmission_interrupt for setup_tx
    report.expect_read(CAN0.n()[0].tx().txbtiei().addr(), 4, 0b0);
    report.expect_read(CAN0.n()[0].tx().txbtiei().addr(), 4, 0b1);

    // txefc0 for set_tx_event_fifo_start_address for setup_tx
    report.expect_read(CAN0.n()[0].tx().txefci().addr(), 4, 0b0);
    // txefc0 for set_tx_event_fifo_size for setup_tx
    report.expect_read(CAN0.n()[0].tx().txefci().addr(), 4, 0b100_0000_0000);

    // cccr0 for set_frame_mode for setup_tx
    report.expect_read(CAN0.n()[0].cccri().addr(), 4, 0b11);

    node.setup_tx(&TxConfig {
        mode: TxMode::DedicatedBuffers,
        dedicated_tx_buffers_number: 2,
        fifo_queue_size: 0,
        buffer_data_field_size: DataFieldSize::_8,
        event_fifo_size: 1,
        tx_event_fifo_start_address: 0x400,
        tx_buffers_start_address: 0x440,
    });

    // rxesc0 for set_rx_buffer_data_field_size for setup_rx
    report.expect_read(CAN0.n()[0].rx().rxesci().addr(), 4, 0b0);

    // rxbc0 for set_rx_buffer_start_address for setup_rx
    report.expect_read(CAN0.n()[0].rx().rxbci().addr(), 4, 0b0);

    // rxesc0 for set_rx_fifo0_data_field_size for setup_rx
    report.expect_read(CAN0.n()[0].rx().rxesci().addr(), 4, 0b0);

    // rxf0c0 for set_rx_fifo0_start_address
    report.expect_read(CAN0.n()[0].rx().rxf0ci().addr(), 4, 0b0);

    // rxf0c0 for set_rx_fifo0_size
    report.expect_read(CAN0.n()[0].rx().rxf0ci().addr(), 4, 0b1_0000_0000);

    // rxf0c0 for set_rx_fifo0_operating_mode
    report.expect_read(CAN0.n()[0].rx().rxf0ci().addr(), 4, 0b0100_0000_0001_0000_0000);

    // rxf0c0 for set_rx_fifo0_watermark_level
    report.expect_read(CAN0.n()[0].rx().rxf0ci().addr(), 4, 0b0100_0000_0001_0000_0000);

    // cccr0 for set_frame_mode for setup_tx
    report.expect_read(CAN0.n()[0].cccri().addr(), 4, 0b11);


    node.setup_rx(RxConfig {
        mode: RxMode::SharedFifo0,
        buffer_data_field_size: DataFieldSize::_8,
        fifo0_data_field_size: DataFieldSize::_8,
        fifo1_data_field_size: DataFieldSize::_8,
        fifo0_operating_mode: RxFifoMode::Blocking,
        fifo1_operating_mode: RxFifoMode::Blocking,
        fifo0_watermark_level: 0,
        fifo1_watermark_level: 0,
        fifo0_size: 4,
        fifo1_size: 0,
        rx_fifo0_start_address: 0x100,
        rx_fifo1_start_address: 0x200,
        rx_buffers_start_address: 0x300,
    });

    // wtdcpu0con0 for clear_cpu_endinit
    report.expect_read(SCU.wdtcpu0con0().addr(), 4, 0b1111111111111100_00000000111100_1_1);
    report.expect_read(SCU.wdtcpu0con0().addr(), 4, 0b1111111111111100_00000000111100_1_1);
    report.expect_read(SCU.wdtcpu0con0().addr(), 4, 0b1111111111111100_00000000111100_1_1);
    // wtdcpu0con0 write
    // wtdcpu0con0 read
    report.expect_read(SCU.wdtcpu0con0().addr(), 4, 0b1111111111111100_00000000111100_0_1);

    // wtdcpu0con0 for set_cpu_endinit
    report.expect_read(SCU.wdtcpu0con0().addr(), 4, 0b1111111111111100_00000000111100_1_0);
    report.expect_read(SCU.wdtcpu0con0().addr(), 4, 0b1111111111111100_00000000111100_1_0);
    report.expect_read(SCU.wdtcpu0con0().addr(), 4, 0b1111111111111100_00000000111100_1_0);

    // wtdcpu0con0 write
    report.expect_read(SCU.wdtcpu0con0().addr(), 4, 0b1111111111111100_00000000111100_0_1);

    // wtdcpu0con0 write
    report.expect_read(SCU.wdtcpu0con0().addr(), 4, 0b1111111111111100_00000000111100_1_1);

    report.expect_read(CAN0.n()[0].npcri().addr(), 4, 0b0);

    // wtdcpu0con0 for clear_cpu_endinit
    report.expect_read(SCU.wdtcpu0con0().addr(), 4, 0b1111111111111100_00000000111100_1_1);
    report.expect_read(SCU.wdtcpu0con0().addr(), 4, 0b1111111111111100_00000000111100_1_1);
    report.expect_read(SCU.wdtcpu0con0().addr(), 4, 0b1111111111111100_00000000111100_1_1);
    // wtdcpu0con0 write
    // wtdcpu0con0 read
    report.expect_read(SCU.wdtcpu0con0().addr(), 4, 0b1111111111111100_00000000111100_0_1);

    // wtdcpu0con0 for set_cpu_endinit
    report.expect_read(SCU.wdtcpu0con0().addr(), 4, 0b1111111111111100_00000000111100_1_0);
    report.expect_read(SCU.wdtcpu0con0().addr(), 4, 0b1111111111111100_00000000111100_1_0);
    report.expect_read(SCU.wdtcpu0con0().addr(), 4, 0b1111111111111100_00000000111100_1_0);

    // wtdcpu0con0 write
    report.expect_read(SCU.wdtcpu0con0().addr(), 4, 0b1111111111111100_00000000111100_0_1);

    // wtdcpu0con0 write
    report.expect_read(SCU.wdtcpu0con0().addr(), 4, 0b1111111111111100_00000000111100_1_1);

    node.setup_pins(&Pins {
        tx: PIN_TX_0_0_P20_8,
        rx: PIN_RX_0_0_P20_7,
    });

    insta::assert_snapshot!(report.take_log());
}
