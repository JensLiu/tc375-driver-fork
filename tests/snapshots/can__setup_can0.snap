---
source: tests/can.rs
assertion_line: 261
expression: report.take_log()
---
r    0xF003624C 04 0x00000003
w    0xF003624C 04 0x000000FD
w    0xF003624C 04 0x000000FE
ldms 0xF0208000 0x00000001 0x00000002
r    0xF0208000 04 0x00000000
r    0xF003624C 04 0x00000002
w    0xF003624C 04 0x000000FD
w    0xF003624C 04 0x000000FF
r    0xF0208030 04 0x00000000
w    0xF0208030 04 0xC0000000
w    0xF0208030 04 0xC0000003
w    0xF0208030 04 0x00000003
r    0xF0208030 04 0x00000003
r    0xF0208218 04 0x00000001
r    0xF0208218 04 0x00000001
w    0xF0208218 04 0x00000001
r    0xF0208218 04 0x00000001
r    0xF0208218 04 0x00000001
w    0xF0208218 04 0x00000000
r    0xF0208218 04 0x00000000
r    0xF0208218 04 0x00000000
w    0xF0208218 04 0x00000001
r    0xF0208218 04 0x00000001
r    0xF0208218 04 0x00000001
w    0xF0208218 04 0x00000003
r    0xF0036034 04 0x21110212
r    0xF0036030 04 0x17230113
r    0xF0036018 04 0x40013A00
r    0xF0036028 04 0x00013F00
r    0xF003602C 04 0x00000101
r    0xF0036034 04 0x21110212
r    0xF020821C 04 0x06000A03
w    0xF020821C 04 0x00030E03
r    0xF02082C8 04 0x00000000
w    0xF02082C8 04 0x00000000
r    0xF02082C0 04 0x00000000
w    0xF02082C0 04 0x00000440
r    0xF02082C0 04 0x00000440
w    0xF02082C0 04 0x00020440
r    0xF02082E0 04 0x00000000
w    0xF02082E0 04 0x00000001
r    0xF02082E0 04 0x00000001
w    0xF02082E0 04 0x00000003
r    0xF02082F0 04 0x00000000
w    0xF02082F0 04 0x00000400
r    0xF02082F0 04 0x00000400
w    0xF02082F0 04 0x00010400
r    0xF0208218 04 0x00000003
w    0xF0208218 04 0x00000003
r    0xF02082BC 04 0x00000000
w    0xF02082BC 04 0x00000000
r    0xF02082AC 04 0x00000000
w    0xF02082AC 04 0x00000300
r    0xF02082BC 04 0x00000000
w    0xF02082BC 04 0x00000000
r    0xF02082A0 04 0x00000000
w    0xF02082A0 04 0x00000100
r    0xF02082A0 04 0x00000100
w    0xF02082A0 04 0x00040100
r    0xF02082A0 04 0x00040100
w    0xF02082A0 04 0x00040100
r    0xF02082A0 04 0x00040100
w    0xF02082A0 04 0x00040100
r    0xF0208218 04 0x00000003
w    0xF0208218 04 0x00000003
ldms 0xF003B414 0xFF000000 0x10000000
r    0xF003624C 04 0x00000003
w    0xF003624C 04 0x000000FD
w    0xF003624C 04 0x000000FE
ldms 0xF003B440 0xF0000000 0x20000000
r    0xF003624C 04 0x00000002
w    0xF003624C 04 0x000000FD
w    0xF003624C 04 0x000000FF
r    0xF0208140 04 0x00000000
w    0xF0208140 04 0x00000001
ldms 0xF003B418 0x000000FF 0x000000A8
r    0xF003624C 04 0x00000003
w    0xF003624C 04 0x000000FD
w    0xF003624C 04 0x000000FE
ldms 0xF003B444 0x0000000F 0x00000002
r    0xF003624C 04 0x00000002
w    0xF003624C 04 0x000000FD
w    0xF003624C 04 0x000000FF
w    0xF003B404 04 0x01000000
r    0xF0208118 04 0x00000000
w    0xF0208118 04 0x00010000
r    0xF00385B4 04 0x00000000
w    0xF00385B4 04 0x00000002
r    0xF00385B4 04 0x00000000
w    0xF00385B4 04 0x02000000
r    0xF00385B4 04 0x00000000
w    0xF00385B4 04 0x00000400
r    0xF0208254 04 0x00000000
w    0xF0208254 04 0x00000001
r    0xF0208218 04 0x00000000
w    0xF0208218 04 0x00000000
r    0xF0208218 04 0x00000000
r    0xF0208218 04 0x00000000
w    0xF0208218 04 0x00000000
r    0xF0208218 04 0x00000000
